_posts/2023-05-09-introducing-timelines-part-1.md

```diff
@@ -0,0 +1,169 @@
+For example, at any point in time we might ask “what is the sum of all purchases?”.
+Aggregations over timelines are cumulative -- as events are observed, the answer to the question changes.
+We call these continuous timelines because each value continues until the next change.
+
+![A plot showing time on the x-axis and values on the y-axis, with steps showing continuous values.][continuous]
+
+Compared to SQL, timelines introduce two requirements -- ordering by time and grouping by entity.
+While the SQL _relation_ -- an unordered multiset or bag -- is useful for unordered data, the additional requirements of timelines make them ideal for reasoning about cause-and-effect.
+Timelines are to temporal data what relations are to static data.
+
+Adding these requirements mean that timelines are not a fit for every data processing task.
+Instead, they allow timelines to be a _better_ fit for data processing tasks that work with events and time.
+In fact, most event streams (eg., Apache Kafka, Apache Pulsar, AWS Kinesis, etc.) provide ordering and partitioning by key.
+
+When thinking about events and time, you likely already picture something like a timeline.
+By matching the way you already think about time, timelines simplify reasoning about events and time.
+By building in the time and ordering requirements, the timeline abstraction allows temporal queries to intuitively express cause and effect.
+
+# Using Timelines for Temporal Queries
+Timelines are the abstraction used in Kaskada for building temporal queries, but data starts and ends outside of Kaskada.
+It is important to understand the flow of data from input, to timelines, and finally to output.
+
+![A diagram showing the lifecycle of temporal queries. Input is processed by a query as discrete and continuous timelines. Output is produced from the resulting timeline][lifecycle]
+
+Every query starts from one or more sources of input data.
+Each input -- whether it is events arriving in a stream or stored in a table, or facts stored in a table -- can be converted to a timeline without losing important context such as the time of each event.
+
+The query itself is expressed as a sequence of operations.
+Each operation creates a timeline from timelines.
+The result of final operation is used as the result of the query.
+Thus, the query produces a timeline.
+That timeline may be discrete or continuous.
+
+The result of a query is a timeline, which may be output to a sink.
+The rows written to the sink may be a history reflecting the changes within a timeline, or a snapshot reflecting the values at a specific point-in-time.
+
+## Inputting Timelines
+Before a query is performed, each input is mapped to a timeline.
+Every input -- whether events from a stream or table or facts in a table -- can be mapped to a timeline without losing the important temporal information, such as when events happened.
+Events become discrete timelines -- values happening at the time of each event.
```

Comment:
Hm, I'm not sure why values "happening" sounds weird to me. It makes sense if I think about it. But you could also do "Events become discrete timelines, where values happen/exist/occur at the time of each event."

You can probably remove the "such as when events happened" in the previous sentence, since it's reiterated here then?